test/test_autodiff.cpp

```diff
@@ -0,0 +1,1093 @@
+}
+
+// Standard normal cumulative distribution function
+template<typename X>
+X Phi(const X& x)
+{
+  return 0.5*erfc(-boost::math::constants::one_div_root_two<double>()*x);
+}
+
+enum CP { call, put };
+
+// Assume zero annual dividend yield (q=0).
+template<typename Price,typename Sigma,typename Tau,typename Rate>
+boost::math::differentiation::autodiff::promote<Price,Sigma,Tau,Rate>
+    black_scholes_option_price(CP cp, double K, const Price& S, const Sigma& sigma, const Tau& tau, const Rate& r)
+{
+  using namespace std;
+  const auto d1 = (log(S/K) + (r+sigma*sigma/2)*tau) / (sigma*sqrt(tau));
+  const auto d2 = (log(S/K) + (r-sigma*sigma/2)*tau) / (sigma*sqrt(tau));
+  static_assert(std::is_same<decltype(S*Phi(d1) - exp(-r*tau)*K*Phi(d2)),
+    decltype(exp(-r*tau)*K*Phi(-d2) - S*Phi(-d1))>::value, "decltype(call) != decltype(put)");
+  if (cp == call)
+    return S*Phi(d1) - exp(-r*tau)*K*Phi(d2);
+  else
+    return exp(-r*tau)*K*Phi(-d2) - S*Phi(-d1);
+}
+
+BOOST_AUTO_TEST_SUITE(test_autodiff)
+
+BOOST_AUTO_TEST_CASE(constructors)
+{
+    constexpr int m = 3;
+    constexpr int n = 4;
+    // Verify value-initialized instance has all 0 entries.
+    const boost::math::differentiation::autodiff::variable<double,m> empty1 = boost::math::differentiation::autodiff::variable<double,m>();
+    for (int i=0 ; i<=m ; ++i)
+        BOOST_REQUIRE(empty1.derivative(i) == 0.0);
+    const auto empty2 = boost::math::differentiation::autodiff::variable<double,m,n>();
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            BOOST_REQUIRE(empty2.derivative(i,j) == 0.0);
+    // Single variable
+    constexpr double cx = 10.0;
+    const auto x = boost::math::differentiation::autodiff::variable<double,m>(cx);
+    for (int i=0 ; i<=m ; ++i)
+        if (i==0)
+            BOOST_REQUIRE(x.derivative(i) == cx);
+        else if (i==1)
+            BOOST_REQUIRE(x.derivative(i) == 1.0);
+        else
+            BOOST_REQUIRE(x.derivative(i) == 0.0);
+    const boost::math::differentiation::autodiff::variable<double,n> xn = x;
+    for (int i=0 ; i<=n ; ++i)
+        if (i==0)
+            BOOST_REQUIRE(xn.derivative(i) == cx);
+        else if (i==1)
+            BOOST_REQUIRE(xn.derivative(i) == 1.0);
+        else
+            BOOST_REQUIRE(xn.derivative(i) == 0.0);
+    // Second independent variable
+    constexpr double cy = 100.0;
+    const auto y = boost::math::differentiation::autodiff::variable<double,m,n>(cy);
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
+                BOOST_REQUIRE(y.derivative(i,j) == cy);
+            else if (i==0 && j==1)
+                BOOST_REQUIRE(y.derivative(i,j) == 1.0);
+            else
+                BOOST_REQUIRE(y.derivative(i,j) == 0.0);
+}
+
+BOOST_AUTO_TEST_CASE(assignment)
+{
+    constexpr int m = 3;
+    constexpr int n = 4;
+    constexpr double cx = 10.0;
```

Comment:
Generally I template my test cases on a Real type, then test them at float, double, long double, and cpp_bin_float_50 precision.

Reply:
### Ambiguous Overload for 10 Binary Operators

I'm encountering an `error: ambiguous overload for ‘operator+’` when compiling
```c++
mp + ad
```
where the LHS `mp` is of type `cpp_bin_float<>` and the RHS `ad` is of type `autodiff_fvar<>`. The two competing candidates are:
* [operator + (const number<B, et_off>& a, const V& b)](https://github.com/boostorg/multiprecision/blob/f3e4cee0ccff02ffbc1b56a06ed693d1029b1d4b/include/boost/multiprecision/detail/no_et_ops.hpp#L53) in boost/multiprecision/detail/no_et_ops.hpp
* [operator+(const typename fvar<RealType,Order>::root_type& ca, const fvar<RealType,Order>& cr)](https://github.com/pulver/autodiff/blob/85d34700537a530472e8b22970b658dfd49065af/include/boost/math/differentiation/autodiff.hpp#L1098) which is a friend function in autodiff.hpp.

This does not happen for `ad + mp` since that is calling `fvar operator+(const root_type&) const;` which is a member of `fvar`.

This problem exists for all 4 arithmetic operators (+,-,*,/) and all 6 comparision operators (==,!=,<,<=,>,>=).

Interestingly, there is **no problem** when using **`cpp_dec_float<>`** under the same conditions.

Looking through the mailing list, it seems like this isn't a particularly new issue. Is there a known/recommended way to resolve this?

Reply:
That's strange - the overloads should be enabled/disabled the same for both types.

In any case the multiprecision ones are only enabled when

```
template <class T, class Num>
struct is_compatible_arithmetic_type
   : public mpl::bool_<
         is_convertible<T, Num>::value
         && !is_same<T, Num>::value
         && !is_number_expression<T>::value>
{};
```

evaluates to `true`, which would imply that `autodiff_fvar<>` is implicitly convertible to `cpp_bin_float`, if that's the case then we have a genuine ambiguity because the conversion could go either way?

Specializing `is_compatible_arithmetic_type` would sort things out in short order, but I'd rather understand why there's an ambiguity first.

Reply:
> ... which would imply that `autodiff_fvar<>` is implicitly convertible to `cpp_bin_float` ...

Yes, in the class definition there is the [declaration](https://github.com/pulver/autodiff/blob/master/include/boost/math/differentiation/autodiff.hpp#L758)
```
explicit operator root_type() const;
```
where `root_type` is an alias to `cpp_bin_float_50` for instance in the case of `autodiff_fvar<cpp_bin_float_50,3> ad`. This operator is invoked in `static_cast<cpp_bin_float_50>(ad)` for example to get back the scalar value, or equivalently `ad.derivative(0)`. This seems to make sense, but am open to a different perspective on whether or not this makes the most sense when dealing with other number types.

Reply:
OK that's weird, it may take a day or two, but let me look into this.

Reply:
This sucks, but it was part of the inherent design of cpp_dec_float - long story short it calculates the inverse and multiplies for division as this is faster than the alternatives, but doesn't give an exact result without correct rounding, and cpp_dec_float does no rounding.... instead it has "guard digits" rather like GMP's `mpf_t`.   `cpp_bin_float` does do exact rounding though.

Reply:
> it calculates the inverse and multiplies for division as this is faster than the alternatives, but doesn't give an exact result

I did exactly this with an earlier implementation of `operator/` overload for autodiff. When I found I was getting similar results above, I decided that having accurate division outweighed the slight performance hit. In the end, the improvement in accuracy was noticable; the difference in speed was not.

Reply:
Tests with `cpp_dec_bin` are omitted due to imprecise arithmetic.